service/app/services/GoogleAuthService.scala

```diff
@@ -0,0 +1,68 @@
+package services
+
+import javax.inject.Inject
+
+import domain.{User, UserRole}
+import google_auth.GoogleAuthResponse
+import hateoas.JsonApi._
+import org.apache.commons.lang3.StringUtils
+import org.joda.time.DateTime
+import play.api.Configuration
+import play.api.libs.ws.WSClient
+import repositories.UserRepository
+import services.GoogleAuthService._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class GoogleAuthService @Inject()(ws: WSClient,
+                                  config: Configuration,
+                                  userRepository: UserRepository)(implicit ex: ExecutionContext) {
+
+  private val clientID = config.getString("google.client-id")
+
+  def authenticate(idToken: String): Future[User] =
+    validateToken(idToken).flatMap { res =>
+      userRepository.findByGoogleId(res.sub).flatMap {
+        case Some(u) => Future.successful(u)
+        case _ =>
+          val userToRegister = User(
+            username = res.email,
+            password = StringUtils.EMPTY,
+            firstName = res.given_name,
+            lastName = res.family_name,
+            role = UserRole.CUSTOMER,
+            address = None,
```

Comment:
Can we obtain the address? Is this important to our **delivery feature**?

Reply:
It's inconvenient to use real-time location for address because we need to deliver order to it. Well, we have address field in the database. I didn't show it in web profile section because I couldn't find any use case for that. Now, if we want to support push notifications for **delivery package nearby feature**, we need to now user's address, right?

Reply:
Of course, but if we won't be able to get the address using Google API, we just will prohibit the user to use that feature until he/she provide us real address.